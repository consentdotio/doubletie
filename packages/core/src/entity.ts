// import type { StandardSchemaV1 } from '@standard-schema/spec';
// import type { DatabaseConfig } from './config';
// import { mergeSchemaWithConfig } from './merge';
// import type { EntitySchemaDefinition, SchemaField } from './schema';
// import type {
// 	EntityFieldReference,
// 	EntityFields,
// 	EntityFromDefinition,
// 	EntityStructure,
// 	GetValidatedRelationshipConfig,
// 	JoinTableConfig,
// 	RelationshipConfig,
// 	ValidateRelationshipConfig,
// 	ValidateRelationships,
// 	ValidatedRelationship,
// } from './types';
// import { validateEntity, validateEntityWithFieldValidators } from './validate';

// // Relationship helpers interface for strong typing
// export interface RelationshipHelpers {
// 	/**
// 	 * Create a type-safe many-to-one relationship to another entity's field
// 	 * @param targetEntity The entity being referenced
// 	 * @param fieldName A field name from the target entity
// 	 * @param options Optional configuration for the relationship
// 	 * @returns A relationship object to be used in field definitions
// 	 *
// 	 * @example
// 	 * ```typescript
// 	 * withRelationships(rel => ({
// 	 *   // Many-to-one relationship (default)
// 	 *   authorId: rel.manyToOne(authorEntity, 'id')
// 	 * }))
// 	 * ```
// 	 */
// 	manyToOne<
// 		E extends ReturnType<typeof defineEntity>,
// 		F extends EntityFields<E> & string,
// 		SourceEntity extends EntityStructure,
// 	>(
// 		targetEntity: E,
// 		fieldName: F,
// 		options?: Omit<
// 			RelationshipConfig<SourceEntity, E>,
// 			'type' | 'foreignKey' | 'joinTable'
// 		>
// 	): ValidatedRelationship<E, F>;

// 	/**
// 	 * Create a type-safe one-to-one relationship to another entity's field
// 	 * @param targetEntity The entity being referenced
// 	 * @param fieldName A field name from the target entity
// 	 * @param options Configuration for the relationship
// 	 * @returns A relationship object to be used in field definitions
// 	 *
// 	 * @example
// 	 * ```typescript
// 	 * // Direct relationship (where this entity has the foreign key)
// 	 * profile.withRelationships(rel => ({
// 	 *   userId: rel.oneToOne(userEntity, 'id')
// 	 * }))
// 	 *
// 	 * // Inverse relationship (where the target entity has the foreign key)
// 	 * user.withRelationships(rel => ({
// 	 *   profile: rel.oneToOne(profileEntity, 'id', {
// 	 *     foreignKey: 'userId'
// 	 *   })
// 	 * }))
// 	 * ```
// 	 */
// 	oneToOne<
// 		E extends ReturnType<typeof defineEntity>,
// 		F extends EntityFields<E> & string,
// 		SourceEntity extends EntityStructure,
// 	>(
// 		targetEntity: E,
// 		fieldName: F,
// 		options?: Omit<
// 			RelationshipConfig<SourceEntity, E>,
// 			'type' | 'joinTable'
// 		> & {
// 			/**
// 			 * For inverse relationships, the foreign key in the target entity
// 			 * that refers to this entity's primary key
// 			 */
// 			foreignKey?: keyof E['fields'] & string;
// 		}
// 	): ValidatedRelationship<E, F>;

// 	/**
// 	 * Create a type-safe one-to-many relationship to another entity
// 	 * @param targetEntity The entity collection being referenced
// 	 * @param fieldName The primary key field of the target entity (usually 'id')
// 	 * @param options Configuration for the relationship
// 	 * @returns A relationship object to be used in field definitions
// 	 *
// 	 * @example
// 	 * ```typescript
// 	 * // One-to-many relationship
// 	 * author.withRelationships(rel => ({
// 	 *   books: rel.oneToMany(bookEntity, 'id', {
// 	 *     foreignKey: 'authorId', // Field in bookEntity that refers to author
// 	 *     cascade: true
// 	 *   })
// 	 * }))
// 	 * ```
// 	 */
// 	oneToMany<
// 		E extends ReturnType<typeof defineEntity>,
// 		F extends EntityFields<E> & string,
// 		SourceEntity extends EntityStructure,
// 	>(
// 		targetEntity: E,
// 		fieldName: F,
// 		options: Omit<RelationshipConfig<SourceEntity, E>, 'type' | 'joinTable'> & {
// 			/**
// 			 * The foreign key in the target entity that refers to this entity's primary key
// 			 * This must be a valid field in the target entity
// 			 */
// 			foreignKey: keyof E['fields'] & string;
// 		}
// 	): ValidatedRelationship<E, F>;

// 	/**
// 	 * Create a type-safe many-to-many relationship with another entity
// 	 * @param targetEntity The entity to create a many-to-many relationship with
// 	 * @param joinTableConfig Configuration for the join table
// 	 * @returns A relationship object to be used in field definitions
// 	 *
// 	 * @example
// 	 * ```typescript
// 	 * withRelationships(rel => ({
// 	 *   // Many-to-many relationship
// 	 *   tags: rel.manyToMany(tagEntity, {
// 	 *     tableName: 'post_tags',
// 	 *     sourceColumn: 'postId',
// 	 *     targetColumn: 'tagId',
// 	 *     additionalColumns: {
// 	 *       addedAt: field('date', { required: true })
// 	 *     }
// 	 *   })
// 	 * }))
// 	 * ```
// 	 */
// 	manyToMany<
// 		E extends ReturnType<typeof defineEntity>,
// 		SourceEntity extends EntityStructure,
// 	>(
// 		targetEntity: E,
// 		joinTableConfig: JoinTableConfig<SourceEntity, E> & {
// 			/**
// 			 * Additional options for the relationship
// 			 */
// 			cascade?: boolean;
// 			fetch?: 'lazy' | 'eager';
// 		}
// 	): {
// 		model: E['name'];
// 		field: 'id'; // Defaulting to id since it's conceptually a collection
// 		relationship: RelationshipConfig<SourceEntity, E>;
// 	};
// }

// // Generate a table using schema information
// function generateTable(entityDef: any) {
// 	// ... This would contain your actual table generation logic
// 	return {} as any; // Placeholder
// }

// export function defineEntity<
// 	TName extends string,
// 	TFields extends Record<string, SchemaField<any, any>>,
// 	TValidator extends StandardSchemaV1,
// >(
// 	schema: { name: TName; fields: TFields },
// 	validator?: TValidator
// ): EntityFromDefinition<TName, TFields, TValidator> {
// 	// Create the base entity with schema fields
// 	const baseEntity = {
// 		name: schema.name,
// 		prefix: '',
// 		fields: { ...schema.fields },
// 		config: {},
// 		order: 0,
// 		validator,
// 	};

// 	function resolveWithConfig(runtimeConfig: DatabaseConfig) {
// 		return mergeSchemaWithConfig(baseEntity, runtimeConfig);
// 	}

// 	// Pre-resolve if config provided
// 	const resolved = config ? resolveWithConfig(config) : undefined;

// 	// Create the entity object with all its methods
// 	const entity = {
// 		...baseEntity,
// 		resolved,
// 		resolveWithConfig,

// 		// Validate using Standard Schema
// 		validate: async (data: StandardSchemaV1.InferInput<V>) => {
// 			if (validator) {
// 				return await validateEntity(validator, data);
// 			} else if (baseEntity.validator) {
// 				return await validateEntity(baseEntity.validator, data);
// 			} else {
// 				// Fall back to field-by-field validation
// 				return await validateEntityWithFieldValidators(
// 					baseEntity,
// 					data as Record<string, unknown>
// 				);
// 			}
// 		},

// 		// Type helpers
// 		type: {} as StandardSchemaV1.InferInput<V>,
// 		outputType: {} as StandardSchemaV1.InferOutput<V>,

// 		// Table generation
// 		getTable: (runtimeConfig?: DatabaseConfig) => {
// 			const entityDef =
// 				resolved ||
// 				(runtimeConfig ? resolveWithConfig(runtimeConfig) : baseEntity);
// 			return generateTable(entityDef as any);
// 		},

// 		// Builder pattern for type-safe relationships
// 		withRelationships<R extends Record<string, EntityFieldReference<any, any>>>(
// 			relationshipFn: (helpers: RelationshipHelpers) => R
// 		) {
// 			// Calculate entity structure type for validation
// 			type ThisEntityStructure = EntityStructure<T['name'], T['fields']>;

// 			// Static type validation of relationships
// 			type ValidatedRels = ValidateRelationships<ThisEntityStructure, R>;

// 			// Ensure ValidatedRels doesn't contain any 'never' types
// 			// This will cause a compile error if invalid relationships are detected
// 			type _EnsureValidRelationships = [ValidatedRels] extends [R]
// 				? true
// 				: false;

// 			// Create relationship helpers with type inference capabilities
// 			const helpers: RelationshipHelpers = {
// 				manyToOne<
// 					TargetE extends ReturnType<typeof defineEntity>,
// 					F extends EntityFields<TargetE> & string,
// 					Config extends RelationshipConfig<
// 						ThisEntityStructure,
// 						TargetE
// 					> = RelationshipConfig<ThisEntityStructure, TargetE>,
// 				>(
// 					targetEntity: TargetE,
// 					fieldName: F,
// 					options?: Omit<Config, 'type' | 'foreignKey' | 'joinTable'>
// 				): ValidatedRelationship<TargetE, F> {
// 					return {
// 						model: targetEntity.name,
// 						field: fieldName,
// 						relationship: {
// 							...options,
// 							type: 'manyToOne',
// 						},
// 					} as ValidatedRelationship<TargetE, F>;
// 				},

// 				oneToOne<
// 					TargetE extends ReturnType<typeof defineEntity>,
// 					F extends EntityFields<TargetE> & string,
// 					Config extends RelationshipConfig<
// 						ThisEntityStructure,
// 						TargetE
// 					> = RelationshipConfig<ThisEntityStructure, TargetE>,
// 				>(
// 					targetEntity: TargetE,
// 					fieldName: F,
// 					options?: Omit<Config, 'type' | 'joinTable'> & {
// 						foreignKey?: keyof TargetE['fields'] & string;
// 					}
// 				): ValidatedRelationship<TargetE, F> {
// 					// If foreignKey is specified, validate it exists in the target entity
// 					if (
// 						options?.foreignKey &&
// 						!targetEntity.fields[options.foreignKey as string]
// 					) {
// 						throw new Error(
// 							`Relationship Error: Foreign key '${options.foreignKey}' does not exist in entity '${targetEntity.name}'.` +
// 								` Available fields: ${Object.keys(targetEntity.fields).join(', ')}`
// 						);
// 					}

// 					return {
// 						model: targetEntity.name,
// 						field: fieldName,
// 						relationship: {
// 							...options,
// 							type: 'oneToOne',
// 						},
// 					} as ValidatedRelationship<TargetE, F>;
// 				},

// 				oneToMany<
// 					TargetE extends ReturnType<typeof defineEntity>,
// 					F extends EntityFields<TargetE> & string,
// 					Config extends RelationshipConfig<
// 						ThisEntityStructure,
// 						TargetE
// 					> = RelationshipConfig<ThisEntityStructure, TargetE>,
// 				>(
// 					targetEntity: TargetE,
// 					fieldName: F,
// 					options: Omit<Config, 'type' | 'joinTable'> & {
// 						foreignKey: keyof TargetE['fields'] & string;
// 					}
// 				): ValidatedRelationship<TargetE, F> {
// 					// Validate the foreignKey exists in the target entity
// 					if (!targetEntity.fields[options.foreignKey as string]) {
// 						throw new Error(
// 							`Relationship Error: Foreign key '${options.foreignKey}' does not exist in entity '${targetEntity.name}'.` +
// 								` Available fields: ${Object.keys(targetEntity.fields).join(', ')}`
// 						);
// 					}

// 					return {
// 						model: targetEntity.name,
// 						field: fieldName,
// 						relationship: {
// 							...options,
// 							type: 'oneToMany',
// 						},
// 					} as ValidatedRelationship<TargetE, F>;
// 				},

// 				manyToMany<
// 					TargetE extends ReturnType<typeof defineEntity>,
// 					Config extends RelationshipConfig<
// 						ThisEntityStructure,
// 						TargetE
// 					> = RelationshipConfig<ThisEntityStructure, TargetE>,
// 				>(
// 					targetEntity: TargetE,
// 					joinTableConfig: JoinTableConfig<ThisEntityStructure, TargetE> & {
// 						cascade?: boolean;
// 						fetch?: 'lazy' | 'eager';
// 					}
// 				): {
// 					model: TargetE['name'];
// 					field: 'id';
// 					relationship: RelationshipConfig<ThisEntityStructure, TargetE>;
// 				} {
// 					// Validate join table config
// 					if (!joinTableConfig.tableName) {
// 						throw new Error(
// 							'Join table name is required for many-to-many relationships'
// 						);
// 					}

// 					const { cascade, fetch, ...joinTableDetails } = joinTableConfig;

// 					// Create a relationship configuration with many-to-many type and the join table config
// 					const relationshipConfig: RelationshipConfig<
// 						ThisEntityStructure,
// 						TargetE
// 					> = {
// 						type: 'manyToMany',
// 						joinTable: joinTableDetails,
// 						cascade,
// 						fetch,
// 					};

// 					// Return the relationship object
// 					return {
// 						model: targetEntity.name,
// 						field: 'id' as const, // We use 'id' by convention for collections
// 						relationship: relationshipConfig,
// 					};
// 				},
// 			};

// 			// Apply relationships with type checking
// 			const relationships = relationshipFn(helpers);

// 			// Update fields with relationships
// 			for (const [fieldName, relationship] of Object.entries(relationships)) {
// 				if (entity.fields[fieldName]) {
// 					entity.fields[fieldName].relationship = relationship;
// 				} else {
// 					throw new Error(
// 						`Relationship Error: Cannot add relationship to non-existent field '${fieldName}' in entity '${entity.name}'.` +
// 							` Available fields: ${Object.keys(entity.fields).join(', ')}`
// 					);
// 				}
// 			}

// 			// Return new entity with relationships
// 			return {
// 				...entity,
// 				relationships,
// 			};
// 		},
// 	};

// 	return entity as EntityFromDefinition<T, V>;
// }
