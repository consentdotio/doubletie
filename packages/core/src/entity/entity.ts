import type { StandardSchemaV1 } from '@standard-schema/spec';
import type { DatabaseConfig } from '../config/config.types';
import { mergeSchemaWithConfig } from '../db/merge';
import { generateTable } from '../db/table';
import type { EntitySchemaDefinition } from '../schema/schema.types';
import {
	validateEntity,
	validateEntityWithFieldValidators,
} from '../validation/validate';
import { convertValueToFieldType } from '../validation/type-conversion';
import type {
	EntityFromDefinition,
	EntityStructure,
	EntityWithRelationships,
	SchemaFields,
} from './entity.types';
import { createRelationshipHelpers } from './relationship';
import type { RelationshipHelpers } from './relationship.types';
import { EntityFieldsMap } from '../db/adapters';

// Type alias to prevent circular reference issues
type EntityType<
	TSchema extends EntitySchemaDefinition,
	TValidator extends StandardSchemaV1,
> = EntityFromDefinition<TSchema, TValidator>;

/**
 * Define a new entity with validation and relationships
 * @param schema The entity schema definition
 * @param validator Optional schema validator
 * @param config Optional database configuration
 * @returns The entity definition with methods for validation and relationships
 */
export function defineEntity<
	TSchema extends EntitySchemaDefinition,
	TValidator extends StandardSchemaV1 = StandardSchemaV1,
>(
	schema: TSchema,
	validator?: TValidator,
	config?: DatabaseConfig
): EntityFromDefinition<TSchema, TValidator> {
	// Create the base entity structure
	const baseEntity = {
		name: schema.name,
		prefix: '',
		fields: { ...schema.fields },
		config: {},
		order: 0,
		validator,
		description: schema.description,
	};

	// Create a resolved entity based on configuration if provided
	let resolved: ReturnType<typeof mergeSchemaWithConfig> | undefined;

	function resolveWithConfig(runtimeConfig: DatabaseConfig) {
		if (!resolved) {
			resolved = mergeSchemaWithConfig(baseEntity, runtimeConfig);
		}
		return resolved;
	}

	// Create the entity with explicit type annotation to avoid circular reference
	const entity: EntityType<TSchema, TValidator> = {
		...baseEntity,

		// Validate data against the entity schema
		validate: async <TInput extends StandardSchemaV1.InferInput<TValidator>>(
			data: TInput
		): Promise<EntityFieldsMap<SchemaFields<TSchema>>> => {
			const entityDef =
				resolved || (config ? resolveWithConfig(config) : baseEntity);

			// Use field-level validators when there's no explicit validator
			if (!validator) {
				// When no validator is provided, perform field-level validation with proper typing
				// Pass input data and schema, specifying output type will be EntityFieldsMap<SchemaFields<TSchema>>
				const validatedData = await validateEntityWithFieldValidators<
					Record<string, unknown>, 
					EntityFieldsMap<SchemaFields<TSchema>>
				>(
					data as Record<string, unknown>,
					schema
				);
				
				// Return the validated data with proper type
				return validatedData;
			}

			// When using an explicit validator, validate with the schema first
			const result = await validateEntity(validator, data);
			
			// Ensure the data has correct types based on field definitions
			const typedResult: Record<string, unknown> = {};
			
			// Process any required field conversions
			for (const [fieldName, fieldDef] of Object.entries(schema.fields)) {
				// Include all fields from the validated data if they exist
				if (result && typeof result === 'object' && fieldName in result) {
					const fieldValue = (result as Record<string, unknown>)[fieldName];
					// Apply automated type conversions based on field type
					typedResult[fieldName] = convertValueToFieldType(fieldValue, fieldDef, fieldName);
				}
			}
			
			// Return the result with the correct EntityFieldsMap type
			return typedResult as EntityFieldsMap<SchemaFields<TSchema>>;
		},

		// Type references for TypeScript autocompletion
		type: {} as StandardSchemaV1.InferInput<TValidator>,
		outputType: {} as StandardSchemaV1.InferOutput<TValidator>,

		// Generate a table definition - using any to avoid complex type issues
		getTable: (runtimeConfig?: DatabaseConfig) => {
			// If config is provided, use mergeSchemaWithConfig to properly format the entity
			const entityDef =
				resolved ||
				(runtimeConfig ? resolveWithConfig(runtimeConfig) : undefined);

			// This uses 'any' to bypass the complex type issues while maintaining runtime functionality
			return entityDef ? generateTable(entityDef) : ({} as any);
		},

		// Builder pattern for type-safe relationships
		withRelationships: function <
			TRel extends Record<
				string,
				{ entity: string; field: string; relationship?: any }
			>,
		>(
			relationshipFn: (
				helpers: RelationshipHelpers<EntityType<TSchema, TValidator>>
			) => TRel
		): EntityWithRelationships<EntityType<TSchema, TValidator>, TRel> {
			// Create relationship helpers instance
			const selfEntity = this as unknown as EntityStructure;
			const helpers = createRelationshipHelpers(selfEntity);
			const typedHelpers = helpers as RelationshipHelpers<
				EntityType<TSchema, TValidator>
			>;
			const relationships = relationshipFn(typedHelpers);

			// Create entity with relationships
			return {
				...this,
				relationships,
			} as any; // Using 'any' to bypass complex type issues
		},
	};

	return entity;
}
