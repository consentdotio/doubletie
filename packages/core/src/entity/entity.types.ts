import type { StandardSchemaV1 } from '@standard-schema/spec';
import type { DatabaseConfig } from '../config/config.types';
import type { TableDefinition, EntityFieldsMap } from '../db/adapters/adapter';
import type { SchemaField } from '../schema/schema.types';
import type { EntitySchemaDefinition } from '../schema/schema.types';
import type {
	EntityFieldReference,
	RelationshipConfig,
	RelationshipHelpers,
	RelationshipType,
} from './relationship.types';

/**
 * Extract all field names from an entity with their types
 * @template TEntity The entity type to extract fields from
 * @template TFieldType Optional field type filter
 */
export type EntityFields<
	TEntity extends { fields: Record<string, SchemaField<any>> },
	TFieldType extends string = string,
> = {
	[K in keyof TEntity['fields'] &
		string]: TEntity['fields'][K]['type'] extends TFieldType ? K : never;
}[keyof TEntity['fields'] & string];

/**
 * Extract the value type for a given field
 * @template TEntity The entity containing fields
 * @template TFieldName The name of the field to get the type for
 */
export type EntityFieldValue<
	TEntity extends { fields: Record<string, SchemaField<any>> },
	TFieldName extends keyof TEntity['fields'] & string,
> = TEntity['fields'][TFieldName] extends SchemaField<infer TType, infer TValue>
	? TValue
	: unknown;

/**
 * Type to represent the structure of an entity
 * @template TName The entity name type
 * @template TFields The entity fields record type
 */
export interface EntityStructure<
	TName extends string = string,
	TFields extends Record<string, SchemaField<any>> = Record<
		string,
		SchemaField<any>
	>,
> {
	name: TName;
	fields: TFields;
	description?: string;
}

/**
 * Type-safe relationship reference - alias for EntityFieldReference for backward compatibility
 * @template TEntity Target entity type
 * @template TFieldName Field name in target entity
 */
export type TypedRelationshipReference<
	TEntity extends EntityStructure,
	TFieldName extends keyof TEntity['fields'] &
		string = keyof TEntity['fields'] & string,
> = EntityFieldReference<EntityStructure, TEntity, TFieldName>;

/**
 * Entity result type after applying relationships
 * @template TEntity The base entity type
 * @template TRelationships Record of relationship references
 */
export type EntityWithRelationships<
	TEntity extends EntityStructure,
	TRelationships extends Record<string, TypedRelationshipReference<any>>,
> = TEntity & {
	relationships: TRelationships;
};

/**
 * Extracts the name type from a schema definition
 */
export type SchemaName<TSchema extends EntitySchemaDefinition> =
	TSchema['name'] extends string ? TSchema['name'] : string;

/**
 * Extracts the fields type from a schema definition
 */
export type SchemaFields<TSchema extends EntitySchemaDefinition> =
	TSchema['fields'] extends Record<string, SchemaField<any>>
		? TSchema['fields']
		: Record<string, SchemaField<any>>;

/**
 * Get the full entity type from a schema definition
 * @template TSchema The schema definition type
 * @template TValidator Validator schema type
 */
export interface EntityFromDefinition<
	TSchema extends EntitySchemaDefinition = EntitySchemaDefinition,
	TValidator extends StandardSchemaV1 = StandardSchemaV1,
> extends EntityStructure<SchemaName<TSchema>, SchemaFields<TSchema>> {
	name: SchemaName<TSchema>;
	prefix: string;
	fields: SchemaFields<TSchema>;
	config: Record<string, unknown>;
	order: number;
	validator?: TValidator;

	// Type-safe validation method that properly infers types from field definitions
	validate: <TInput extends StandardSchemaV1.InferInput<TValidator>>(
		data: TInput
	) => Promise<
		EntityFieldsMap<SchemaFields<TSchema>> 
	>;

	// Type references for input/output
	type: StandardSchemaV1.InferInput<TValidator>;
	outputType: StandardSchemaV1.InferOutput<TValidator>;

	// Type-safe table generation
	getTable: <TOptions extends Record<string, any> = {}>(
		config?: DatabaseConfig
	) => TableDefinition<TOptions>;

	// Type-safe relationship builder
	withRelationships: <
		TRel extends Record<string, TypedRelationshipReference<any>>,
	>(
		relationshipFn: (
			helpers: RelationshipHelpers<EntityFromDefinition<TSchema, TValidator>>
		) => TRel
	) => EntityWithRelationships<EntityFromDefinition<TSchema, TValidator>, TRel>;
}

/**
 * Helper type to infer input type from an entity with improved type safety
 * @template TEntity Entity with validate method
 */
export type EntityInput<
	TEntity extends {
		validate: (data: any) => any;
		fields: Record<string, SchemaField<any>>;
	},
> = Parameters<TEntity['validate']>[0];

/**
 * Helper type to infer output type from an entity with improved type safety
 * @template TEntity Entity with validate method
 */
export type EntityOutput<
	TEntity extends {
		validate: (data: any) => Promise<any>;
		fields: Record<string, SchemaField<any>>;
	},
> = Awaited<ReturnType<TEntity['validate']>>;

/**
 * Type to get keys of an entity's fields that match a specific field type
 * @template TEntity The entity to extract fields from
 * @template TFieldType The field type to filter by
 */
export type EntityFieldsByType<
	TEntity extends EntityStructure,
	TFieldType extends string,
> = Extract<keyof TEntity['fields'], EntityFields<TEntity, TFieldType>>;

/**
 * Helper type to infer an entity's relationship capabilities
 * @template TEntity The base entity type
 */
export type WithRelationships<TEntity extends EntityStructure> = TEntity & {
	withRelationships: <
		TRel extends Record<string, TypedRelationshipReference<EntityStructure>>,
	>(
		relationshipFn: (
			helpers: RelationshipHelpers<TEntity | EntityStructure>
		) => TRel
	) => EntityWithRelationships<TEntity, TRel>;
};

/**
 * Extract all the fields from an entity that have a specific property
 * @template TEntity The entity to extract fields from
 * @template TProp The property name to check for
 */
export type EntityFieldsWithProperty<
	TEntity extends EntityStructure,
	TProp extends keyof SchemaField<any>,
> = {
	[K in keyof TEntity['fields'] &
		string]: TProp extends keyof TEntity['fields'][K] ? K : never;
}[keyof TEntity['fields'] & string];

/**
 * Get primary key fields from an entity
 * @template TEntity The entity to extract primary key fields from
 */
export type EntityPrimaryKeys<TEntity extends EntityStructure> =
	EntityFieldsWithProperty<TEntity, 'primaryKey'>;

/**
 * Get required fields from an entity
 * @template TEntity The entity to extract required fields from
 */
export type EntityRequiredFields<TEntity extends EntityStructure> = {
	[K in keyof TEntity['fields'] &
		string]: TEntity['fields'][K]['required'] extends true ? K : never;
}[keyof TEntity['fields'] & string];

/**
 * Create a type-safe input object for an entity
 * @template TEntity The entity to create input type for
 */
export type TypedEntityInput<TEntity extends EntityStructure> = {
	[K in EntityRequiredFields<TEntity>]: EntityFieldValue<TEntity, K>;
} & {
	[K in Exclude<
		keyof TEntity['fields'] & string,
		EntityRequiredFields<TEntity>
	>]?: EntityFieldValue<TEntity, K>;
};
