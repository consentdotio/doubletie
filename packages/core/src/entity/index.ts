/**
 * Entity module exports
 * @module entity
 */

// Entity definition exports
export { defineEntity } from './entity';
export { createRelationshipHelpers } from './relationship';

// Entity types
export type {
	EntityFields,
	EntityStructure,
	EntityFromDefinition,
	EntityWithRelationships,
	EntityInput,
	EntityOutput,
	EntityFieldsByType,
	WithRelationships,
} from './entity.types';

// Relationship types
export type {
	RelationshipType,
	RelationshipConfig,
	EntityFieldReference,
	RelationshipHelpers,
	BasicRelationshipOptions,
	ManyToManyOptions,
} from './relationship.types';
