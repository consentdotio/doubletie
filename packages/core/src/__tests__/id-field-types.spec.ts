import { describe, test, expect, vi, beforeEach } from 'vitest';
import { defineEntity } from '../entity/entity';
import {
    createField,
    idField,
    incrementalIdField,
    prefixedIdField,
    uuidField,
} from '../schema/fields';

// Mock global sequences for consistent test snapshots
vi.mock('uuid', () => ({
    v4: () => '123e4567-e89b-12d3-a456-426614174000'
}));

vi.mock('nanoid', () => ({
    nanoid: () => 'Xc9aEb3Z1f'
}));

describe('ID Field Types', () => {
    beforeEach(() => {
        // Reset sequences for incremental IDs
        vi.restoreAllMocks();
        // Reset custom sequence
        if ((globalThis as any).__sequences) {
            (globalThis as any).__sequences = undefined;
        }
    });

    test('UUID field generates and validates UUIDs', async () => {
        const userEntity = defineEntity({
            name: 'user',
            fields: {
                // UUID: "123e4567-e89b-12d3-a456-426614174000"
                id: uuidField({ required: false }), // Make ID not required for testing
                name: createField('string', { required: true }),
                email: createField('string', { required: true }),
            },
        });

        const user = await userEntity.validate({
            name: 'John Doe',
            email: 'john@example.com',
        });

        // Check that ID was generated and is the mock UUID
        expect(user.id).toBe('123e4567-e89b-12d3-a456-426614174000');
        expect(user).toMatchSnapshot();
    });

    test('Prefixed ID field generates and validates prefixed nanoids', async () => {
        const productEntity = defineEntity({
            name: 'product',
            fields: {
                // Prefixed nanoid: "prod_Xc9aEb3Z1f"
                id: prefixedIdField('prod_', { length: 8, required: false }), // Make ID not required for testing
                name: createField('string', { required: true }),
                price: createField('number', { required: true }),
            },
        });

        const product = await productEntity.validate({
            name: 'Premium Widget',
            price: 49.99,
        });

        // Check that ID was generated and has the correct format
        expect(product.id).toBe('prod_Xc9aEb3Z1f');
        expect(product).toMatchSnapshot();
    });

    test('Incremental ID field generates and validates sequential IDs', async () => {
        const orderEntity = defineEntity({
            name: 'order',
            fields: {
                // Incremental ID: 1000, 1001, 1002...
                id: incrementalIdField(1000, { required: false }), // Make ID not required for testing
                items: createField('array', { required: true }),
            },
        });

        const order1 = await orderEntity.validate({ items: ['item1', 'item2'] });
        const order2 = await orderEntity.validate({ items: ['item3'] });

        // Check that IDs were generated as expected
        expect(order1.id).toBe(1000);
        expect(order2.id).toBe(1001);
        expect(order1).toMatchSnapshot('order1');
        expect(order2).toMatchSnapshot('order2');
    });

    test('Custom ID field generates and validates with custom format', async () => {
        // Mock date to ensure consistent test results
        const mockDate = new Date('2024-01-01');
        vi.setSystemTime(mockDate);

        const ticketEntity = defineEntity({
            name: 'ticket',
            fields: {
                // Custom generator - ticket numbers like "TKT-2024-00001"
                id: idField({
                    idType: 'custom',
                    required: false, // Make ID not required for testing
                    generator: () => {
                        const year = new Date().getFullYear();
                        // For demo purposes only - in production we'd use a database sequence
                        if (!(globalThis as any).__sequences) {
                            (globalThis as any).__sequences = { ticket: 1 };
                        }
                        const num = (globalThis as any).__sequences.ticket++;
                        return `TKT-${year}-${num.toString().padStart(5, '0')}`;
                    },
                }),
                subject: createField('string', { required: true }),
            },
        });

        const ticket1 = await ticketEntity.validate({ subject: 'Technical issue' });
        const ticket2 = await ticketEntity.validate({ subject: 'Feature request' });

        // Check that IDs were generated correctly
        expect(ticket1.id).toBe('TKT-2024-00001');
        expect(ticket2.id).toBe('TKT-2024-00002');
        expect(ticket1).toMatchSnapshot('ticket1');
        expect(ticket2).toMatchSnapshot('ticket2');

        // Reset the mocked date
        vi.useRealTimers();
    });
}); 