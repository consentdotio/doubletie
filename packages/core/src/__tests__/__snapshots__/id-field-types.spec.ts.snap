// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ID Field Types > Custom ID field generates and validates with custom format > ticket1 1`] = `
{
  "id": "TKT-2024-00001",
  "subject": "Technical issue",
}
`;

exports[`ID Field Types > Custom ID field generates and validates with custom format > ticket2 1`] = `
{
  "id": "TKT-2024-00002",
  "subject": "Feature request",
}
`;

exports[`ID Field Types > Incremental ID field generates and validates sequential IDs > order1 1`] = `
{
  "id": 1000,
  "items": [
    "item1",
    "item2",
  ],
}
`;

exports[`ID Field Types > Incremental ID field generates and validates sequential IDs > order2 1`] = `
{
  "id": 1001,
  "items": [
    "item3",
  ],
}
`;
